lib/Conversion/TosaToTTIR/TosaToTTIR.cpp

```diff
@@ -0,0 +1,118 @@
+
+using namespace mlir;
+using namespace tt;
+
+namespace mlir::tt::ttir {
+
+#define GEN_PASS_DEF_CONVERTTOSATOTTIR
+#include "ttmlir/Conversion/Passes.h.inc"
+
+} // namespace mlir::tt::ttir
+
+namespace {
+
+template <typename SrcOp, typename DestOp,
+          typename Adaptor = typename SrcOp::Adaptor>
+class TosaToTTIROpConversionPattern : public OpConversionPattern<SrcOp> {
+  using OpConversionPattern<SrcOp>::OpConversionPattern;
+
+public:
+  LogicalResult
+  matchAndRewrite(SrcOp srcOp, Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if constexpr (std::is_same<SrcOp, tosa::MulOp>::value) {
+      assert(srcOp.getShift() == 0);
+    }
+
+    auto outputType =
+        srcOp.getResult().getType().template cast<RankedTensorType>();
+    auto outputTensor = rewriter.create<tensor::EmptyOp>(
+        srcOp.getLoc(), outputType.getShape(), outputType.getElementType());
+    rewriter.replaceOpWithNewOp<DestOp>(
+        srcOp, TypeRange(outputTensor.getType()), adaptor.getOperands(),
+        ValueRange(outputTensor),
+        rewriter.getArrayAttr(
+            SmallVector<Attribute>(adaptor.getOperands().size() + 1,
+                                   rewriter.getAttr<OperandConstraintAttr>(
+                                       OperandConstraint::AnyDeviceTile))));
+    return success();
+  }
+};
+
+struct ConvertTosaToTTIRPass
+    : public ttir::impl::ConvertTosaToTTIRBase<ConvertTosaToTTIRPass> {
+  void runOnOperation() override {
+    mlir::ConversionTarget target(getContext());
+
+    target.addIllegalDialect<tosa::TosaDialect>();
+
+    target.addLegalDialect<ttir::TTIRDialect>();
+    target.addLegalOp<mlir::tensor::EmptyOp>();
+    target.addLegalOp<mlir::ModuleOp>();
+    target.addLegalOp<mlir::func::FuncOp>();
+    target.addLegalOp<mlir::func::ReturnOp>();
+
+    // For now keep the same type assuming tosa ops operate on builtin tensor.
+    TypeConverter typeConverter;
+    typeConverter.addConversion([](Type type) {
+      assert(type.isa<RankedTensorType>() &&
+             "only ranked tensor type supported");
+      return type;
+    });
+    RewritePatternSet patterns(&getContext());
+
+    // Add conversion patterns.
+    patterns
+        .add<TosaToTTIROpConversionPattern<tosa::AddOp, mlir::tt::ttir::AddOp>>(
+            typeConverter, &getContext());
+    patterns.add<
+        TosaToTTIROpConversionPattern<tosa::MulOp, mlir::tt::ttir::MultiplyOp>>(
+        typeConverter, &getContext());
+    patterns.add<
+        TosaToTTIROpConversionPattern<tosa::SubOp, mlir::tt::ttir::SubtractOp>>(
+        typeConverter, &getContext());
+    patterns.add<TosaToTTIROpConversionPattern<tosa::GreaterEqualOp,
+                                               mlir::tt::ttir::GreaterEqualOp>>(
+        typeConverter, &getContext());
+
+    // Apply conversion.
+    if (failed(
+            applyFullConversion(getOperation(), target, std::move(patterns)))) {
```

Comment:
Have you considered using partial conversion in order to not have to mark built in ops as legal explicitly?

Reply:
TTNN->EmitC perform full conversion because it needs to convert FuncOp, CallOp and ReturnOp explicitly. Most conversion passes in llvm are implemented with partial conversion. Tho being explicit about what's legal makes sense from the current perspective.

Reply:
I guess it makes sense what you're saying. However, even in LLVM project I see many conversions using `full` type.   
We could remove "unneeded" legal ops module, func, return which are kind of needed for our UTs. Or we could omit them from UTs. Having partial conversion may have its drawbacks in the future if someone wants to lower some TOSA module but we don't support some ops in it, compiler would crash somewhere but it would not assert in tosa->ttir conversion pass. That's why I'd prefer `fullConversion`, but maybe I'm wrong as I am new to the MLIR. 

Anyway, I can see reasons for both approaches, so if you don't have hard opinion on it, I would leave it as it is.

Reply:
Ok, i'm fine with leaving as is. I agree that we don't have enough foresight atm to know if one of the other make more sense.